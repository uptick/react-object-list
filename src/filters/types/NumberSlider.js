import React from 'react'
import PropTypes from 'prop-types'

/**
 * Filter input used to pass a range of integers
 * i.e. Greater Than 20
 */
class NumberSlider extends React.Component {
  static propTypes = {
    /** Highest number that can be selected */
    max: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    /** Lowest number that can be selected */
    min: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    /** Current value of filter */
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    /** Function to call when value changed */
    onChange: PropTypes.func,
    /** True if logarithmic value increase enabled */
    logarithmic: PropTypes.bool,
    /** Base to use for logarithmic increase */
    logbase: PropTypes.number,
    /** Decimal places allowed */
    precision: PropTypes.number,
  }

  static defaultProps = {
    value: '',
    min: 0,
    max: 100,
    logarithmic: false,
    logbase: 10,
    precision: 0,
  }

  constructor(props) {
    super(props)
    this.state = {
      ...this.state,
      currentValue: props.value || props.min,
      sliderValue: props.value ? this.getValueForSlider(props.value) : props.min,
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.value !== this.props.value) {
      this.setState({currentValue: nextProps.value})
    }
  }

  /**
   * Handles number input change.
   * Calls onChange with the current value.
   * @param  {event} event onChange event object
   */
  handleValueChange = (event) => {
    const {precision} = this.props
    const value = parseFloat(event.target.value)
    this.setState(
      () => ({
        currentValue: value,
        sliderValue: this.getValueForSlider(value),
      }), () => {
        const roundedValue = value.toFixed(precision)
        this.props.onChange(roundedValue)
      }
    )
  }

  /**
   * Handles slider onMouseUp.
   * Calls onChange with the current translated value.
   * @param  {event} event onMouseUp event object
   */
  handleSliderValueFinalise = (event) => {
    const {precision} = this.props
    const value = parseFloat(event.target.value)
    const finalValue = this.getValueFromSlider(value)
    this.setState(
      () => ({
        currentValue: finalValue,
        sliderValue: value,
      }), () => {
        const roundedValue = finalValue.toFixed(precision)
        this.props.onChange(roundedValue)
      }
    )
  }

  /**
   * Translates slider event value into valid output value
   * If not using logarithmic, this will be the same
   * @param  {number} input Number returned by slider
   * @return {number}       Number stored in state
   */
  getValueFromSlider = input => {
    const {min, max, logarithmic, logbase} = this.props
    let value = parseFloat(input)
    if (logarithmic) {
      const range = max - min
      const fractionAlongSlider = (value - min) / range
      value = (range * ((logbase ** fractionAlongSlider) - 1) / (logbase - 1)) + min
    }
    return value
  }

  /**
   * Translates current value into a slider value
   * If not using logarithmic this will be the same
   * @param  {number} input Actual value
   * @return {number}       Value translated for slider
   */
  getValueForSlider = input => {
    const {min, max, logarithmic, logbase} = this.props
    let value = parseFloat(input)
    if (logarithmic) {
      const logConst = 1 / Math.log(this.props.logbase)
      const range = max - min
      value = min + (range * logConst * Math.log(((value - min) * (logbase - 1) / range) + 1))
    }
    return value
  }

  /**
   * Handles slider change (on mouse up)
   * Sets state to final value
   */
  handleSliderValueChange = (event) => {
    const value = event.target.value
    this.setState(() => ({
      currentValue: this.getValueFromSlider(value),
      sliderValue: value,
    }))
  }

  render() {
    const {min, max, precision} = this.props
    const {currentValue, sliderValue} = this.state
    const step = 1 / (10 ** (precision))
    const parsedValue = parseFloat(currentValue)
    const display = isNaN(parsedValue) ? '' : parsedValue.toFixed(precision)
    const sliderRoundedValue = parseFloat(sliderValue).toFixed(10)
    return (
      <div className="objectlist-row objectlist-current-filter__active-status objectlist-number-slider-filter">
        <input
          type="number"
          className="objectlist-input objectlist-input__number-slider--number"
          min={min}
          max={max}
          value={display}
          onChange={this.handleValueChange}
          step={step}
        />
        <input
          type="range"
          className="objectlist-input__number-slider--slider"
          value={sliderRoundedValue}
          onChange={this.handleSliderValueChange}
          onMouseUp={this.handleSliderValueFinalise}
          step="any"
          min={min}
          max={max}
        />
      </div>
    )
  }
}

export default NumberSlider
