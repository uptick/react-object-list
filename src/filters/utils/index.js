export FilterComparison from './FilterComparison'
export FilterLabel from './FilterLabel'
export RemoveFilter from './RemoveFilter'
export makeFilter from './makeFilter'
export * from './utils'
