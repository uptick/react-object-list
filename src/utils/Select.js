import React from 'react'
import PropTypes from 'prop-types'
import ReactSelect, { components } from 'react-select'
import ReactAsyncSelect from 'react-select/lib/Async'
const {Option} = components

// A searchable set of props to ignore when checking for
// unknown props.
const KNOWN_PROPS = new Set([
  'options', 'value', 'onChange',
  'placeholder', 'className', 'name',
  'filterOption', 'cacheOptions',
])

/**
 * Convert a set of props according to an object of transforms.
 *
 * A helper function to assist in translating props from one format
 * to another. The transforms are an object whose keys are the
 * source prop name, and whose values are an array with two
 * elements. The first is the name of the destination prop, the
 * second is a function that returns the new prop value. The
 * function takes the source prop value.
 *
 * @returns The transformed props.
 */
function convertProps(props, transforms) {
  const nextProps = {}
  for (const [src, [dst, op]] of Object.entries(transforms)) {
    const value = props[src]
    if (value !== undefined) {
      nextProps[dst] = op ? op(value) : value
    }
  }

  // remove props that are not required
  delete props.resetValue
  delete props.required

  // Complain loudly if there are any props that we've not
  // been able to understand.
  for (const attr of Object.keys(props)) {
    if (KNOWN_PROPS.has(attr)) {
      nextProps[attr] = props[attr]
    } else if (!transforms.hasOwnProperty(attr)) {
      throw new Error(
        `Unknown prop during wrapping of react-select v1 to v2: ${attr}`
      )
    }
  }

  return nextProps
}

/**
 * A thin wrapper around react-select.
 *
 * This wrapper makes it easy to shift between versions of
 * react-select. We tended to use react-select in many different
 * places, meaning if the API changed we'd need to update our
 * code everywhere. This way we can keep an internally consistent
 * interface and change this wrapper only.
 *
 * The current mapping defined here is that for converting from
 * react-select v1 to v2.
 */
class Select extends React.Component {
  static propTypes = {
    isAsync: PropTypes.bool,
    optionRenderer: PropTypes.func,
    controlStyle: PropTypes.object,
    menuStyle: PropTypes.object,
    menuContainerStyle: PropTypes.object,
    value: PropTypes.any,
    inputProps: PropTypes.object,
  }

  state = {
    startedTyping: false,
  }

  handleInputChange = inputValue => {
    this.setState({startedTyping: !!inputValue.length})
  }

  render() {
    const {
      isAsync = false,
      controlStyle,
      menuStyle,
      menuContainerStyle,
      optionRenderer,
      inputProps,
      ...otherProps
    } = this.props

    // Convert props semi-automatically.
    const nextProps = convertProps(otherProps, {
      valueKey: ['getOptionValue', v => o => o[v]],
      labelKey: ['getOptionLabel', v => o => o[v]],
      disabled: ['isDisabled'],
      multi: ['isMulti'],
      clearable: ['isClearable'],
      openOnFocus: ['openMenuOnClick'],
      cache: ['cacheOptions'],
      editable: ['isDisabled', editable => !editable],
      autoload: ['defaultOptions'],
      noResultsText: ['noOptionsMessage', text => () => {
        if (this.state.startedTyping) {
          return text
        }
        return 'Start typing to search...'
      }],
      onInputKeyDown: ['onKeyDown'],
      loadOptions: ['loadOptions', fn => (inputValue, callback) => {
        if (callback === undefined) {
          return fn(inputValue) // TODO: check value of this is ok
        } else {
          // This is an ugly hack to ensure we support Promise loading as well as callbacks
          // https://deploy-preview-2289--react-select.netlify.com/async
          // We assume that if the loadOptions function takes 2 arguments, the second is a callback
          if (fn.length === 2) {
            // eslint-disable-next-line handle-callback-err
            fn(inputValue, (err, data) => {
              callback(data.options)
            })
          } else {
            // Assume function takes just one argument, inputValue
            const data = fn(inputValue)
            // Check if the function returned a promise
            if (typeof data.then === 'function') {
              // Wait for result, and call callback
              data.then(({options}) => callback(options))
            } else {
              // Otherwise just call callback straight away
              callback(data.options)
            }
          }
        }
      }],
    })

    // Component overrides are a bit more difficult, and can't be
    // perfectly mapped. We'll need to update our code to work with
    // V2.
    const components = {}
    if (optionRenderer) {
      components.Option = props => <Option {...props}>{optionRenderer(props.data)}</Option>
    }
    if (inputProps) {
      components.innerProps = inputProps
    }
    if (Object.keys(components).length > 0) {
      nextProps.components = components
    }

    // Style overrides are a bit trickier too.
    const styles = {
      control: (base, state) => ({
        ...base,
        backgroundColor: '#fff',
        minHeight: 0,
        ...(controlStyle || {}),
      }),
      dropdownIndicator: (base, state) => ({
        ...base,
        paddingTop: 0,
        paddingBottom: 0,
      }),
      multiValue: (base, state) => ({
        ...base,
        backgroundColor: 'rgba(0,126,255,0.08)',
        border: '1px solid rgba(0,126,255,0.24)',
        color: '#007eff',
      }),
      multiValueLabel: (base, state) => ({
        ...base,
        color: '#007eff',
        borderRight: '1px solid rgba(0,126,255,0.24)',
        fontSize: '90%',
      }),
    }
    if (menuStyle) {
      styles.menu = (base, state) => ({...base, ...menuStyle})
    }
    if (menuContainerStyle) {
      styles.menuList = (base, state) => ({...base, ...menuContainerStyle})
    }
    if (Object.keys(styles).length > 0) {
      nextProps.styles = styles
    }

    const Component = isAsync ? ReactAsyncSelect : ReactSelect
    return (
      <Component {...nextProps} onInputChange={this.handleInputChange} />
    )
  }
}

class AsyncSelect extends React.Component {
  static propTypes = {
    noResultsText: PropTypes.string,
  }

  render() {
    return (
      <Select
        {...this.props}
        noResultsText={this.props.noResultsText || 'No Options'}
        cacheOptions
        isAsync
      />
    )
  }
}

Select.Async = AsyncSelect

export default Select
