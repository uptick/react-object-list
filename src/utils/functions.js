/**
 * Filters columns by those in extraColumns to get the visible or optional fields
 * @param  {object[]} columns      Array of columns or arrays of columns
 * @param  {string[]} extraColumns Array of keys of columns currently visible
 * @param  {bool} optional         The value optional is set to in the particular column
 * @return {object[]}              Returns only those columns which are visible
 */
const getVisibleColumns = (columns, extraColumns = [], optional = false) => {
  return setColumnLabels(columns).reduce((acc, column) => {
    if (column.hasOwnProperty('columns')) {
      // check the next node down
      const visible = column.optional !== undefined && _checkVisibilityHelper(column, extraColumns, optional)
      if (column.optional === undefined || visible) {
        const columns = getVisibleColumns(column.columns, extraColumns, optional)
        if (columns.length || visible) {
          acc.push({...column, columns})
        }
      }
    } else if (Array.isArray(column)) {
      // is a grouped column, check each of the items in the group
      const columns = column.filter(leafColumn => _checkVisibilityHelper(leafColumn, extraColumns, optional))
      if (columns.length) {
        acc.push(columns)
      }
    } else if (_checkVisibilityHelper(column, extraColumns, optional)) {
      // is a leaf node, check individual column
      acc.push(column)
    }
    return acc
  }, [])
}

const _checkVisibilityHelper = (column, extraColumns, optional) => {
  return !!column.optional === optional ||
  extraColumns.includes(column.fieldKey)
}

/** get all the leaves of an unbalanced tree */
const getLeafColumns = (columns) => {
  return columns.reduce((acc, column) => {
    if (column.hasOwnProperty('columns')) {
      return acc.concat(getLeafColumns(column.columns))
    }
    acc.push(column)
    return acc
  }, [])
}

/**
 * Annotates each column with a _colSpan and _rowSpan in the tree
 * which is used when rendering them
 *
 * @param {object[]} columns  Array of columns
 */
const annotateSpans = columns => {
  const depth = _getTreeDepth(columns)
  _annotateColumnSpan(columns, depth)
}

/** depth first search annotating each column with a _colSpan or _rowSpan as appropriate */
const _annotateColumnSpan = (columns, depth) => {
  let sum = 0
  columns.forEach(column => {
    if (!column.hasOwnProperty('columns')) {
      column._colSpan = 1
      column._rowSpan = depth
    } else {
      column._colSpan = _annotateColumnSpan(column.columns, depth - 1)
    }
    sum += column._colSpan
  })
  return sum
}

/** breadth first search for tree depth, as it might be imbalanced */
const _getTreeDepth = (columns) => {
  if (!columns.length) return 0

  let newColumns = []
  columns.forEach(column => {
    if (column.hasOwnProperty('columns')) {
      newColumns = newColumns.concat(column.columns)
    }
  })
  return _getTreeDepth(newColumns) + 1
}

const setColumnLabels = (columns) => {
  return columns.map(column => {
    if (column.hasOwnProperty('columns')) {
      return {
        ...column,
        columns: setColumnLabels(column.columns),
        fieldKey: column.fieldKey || column.header.replace(' ', '').toLowerCase(),
      }
    } else if (Array.isArray(column)) {
      const labels = column.map(({dataKey, fieldKey}) => fieldKey || dataKey.substring(dataKey.lastIndexOf('.') + 1))
      const fieldKeys = column.map(({dataKey, fieldKey}) => fieldKey || dataKey.substring(dataKey.lastIndexOf('.') + 1))
      return column.map(column => {
        const label = column.label || labels.join('_').replace(' ', '').toLowerCase()
        const fieldKey = column.fieldKey || fieldKeys.join('_').replace(' ', '').toLowerCase()
        const labelledColumn = {...column, label, fieldKey}
        labels.push(labels.shift())
        fieldKeys.push(fieldKeys.shift())
        return labelledColumn
      })
    } else {
      return {
        ...column,
        label: (column.fieldKey || column.dataKey.substring(column.dataKey.lastIndexOf('.') + 1)).replace(' ', '').toLowerCase(),
        fieldKey: (column.fieldKey || column.dataKey.substring(column.dataKey.lastIndexOf('.') + 1)),
      }
    }
  })
}

const sortByName = (a, b) => {
  const textA = a.name.toLowerCase()
  const textB = b.name.toLowerCase()
  return ((textA < textB) ? -1 : (textA > textB) ? 1 : 0)
}

export {
  _getTreeDepth,
  getLeafColumns,
  setColumnLabels,
  annotateSpans,
  getVisibleColumns,
  sortByName,
}
