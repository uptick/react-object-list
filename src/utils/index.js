export loadingSpinner from './loading'
export ErrorMessage from './ErrorMessage'

export {
  DATE_FORMAT,
  DATETIME_FORMAT,
  SHORTDATE_FORMAT,
  SHORTDATETIME_FORMAT,
  API_DATE_FORMAT,
  allowedLookupExpressions,
} from './constants'
export {
  COLUMN_TYPE,
  COLUMN_BASE_TYPE,
  FILTER_BASE_TYPE,
  META_TYPE,
  STATUS_TYPE,
  STATUS_CHOICES,
  ALL_SELECTED,
  SELECTION_TYPE,
} from './proptypes'
export {
  getVisibleColumns,
  setColumnLabels,
  sortByName,
  getLeafColumns,
} from './functions'
