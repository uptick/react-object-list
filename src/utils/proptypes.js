import PropTypes from 'prop-types'

const STATUS_CHOICES = {
  loading: 'loading',
  error: 'error',
  done: 'done',
}
const STATUS_TYPE = PropTypes.oneOf(Object.values(STATUS_CHOICES))

const ALL_SELECTED = 'all'
function selectionTypeValidator(isRequired) {
  return function(props, propName, componentName) {
    const value = props[propName]
    if (isRequired && (value === null || value === undefined)) {
      return new Error(`Missing prop \`${propName}\`
      supplied to \`${componentName}\` expected one of ['${ALL_SELECTED}', obj]`)
    }
    let validObject = true
    if (typeof value === 'object') {
      validObject = Object.entries(value).every(
        ([key, value]) => (typeof key === 'number' || 'string') && typeof value === 'boolean')
    } else {
      validObject = false
    }

    if (value !== ALL_SELECTED && !validObject) {
      return new Error(`Invalid prop \`${propName}\` of \`${value}\`
        supplied to \`${componentName}\` expected one of ['${ALL_SELECTED}', obj]`)
    }
  }
}
const SELECTION_TYPE = selectionTypeValidator(false)
SELECTION_TYPE.isRequired = selectionTypeValidator(true)

const COLUMN_BASE_TYPE = {
  fieldKey: PropTypes.string, // used to uniquely identify this column
  dataKey: PropTypes.string, // used to access the data inside the data object
  header: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  item: PropTypes.func,
  sortKey: PropTypes.string,
  optional: PropTypes.bool,
  /* string or function of a data row returning a string to add to row classes */
  rowClass: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
}

const COLUMN_TYPE = PropTypes.shape(COLUMN_BASE_TYPE)

const FILTER_BASE_TYPE = {
  /** Name of filter, used to render filter */
  name: PropTypes.string,
  /** Identifier key for filter */
  filterKey: PropTypes.string,
  /** If true, indicates the filter cannot be removed */
  permanent: PropTypes.bool,
}

const META_TYPE = PropTypes.shape({
  /** the current sortkeys applied using {key: asc_desc_bool } eg. {name: true} for ascending */
  sortKeys: PropTypes.arrayOf(PropTypes.object),
  /** current page highlighted on the pagination */
  currentPage: PropTypes.number,
  /** the maximum number of items shown on a page */
  perPage: PropTypes.number,
  /** the keys of the optional columns that are currently being displayed */
  extraColumns: PropTypes.arrayOf(PropTypes.string),
  /** the total number of items in the dataset if only displaying a subset */
  totalCount: PropTypes.number,
  /** the list of items currently selected */
  selection: PropTypes.array,
})

export {
  COLUMN_TYPE,
  COLUMN_BASE_TYPE,
  FILTER_BASE_TYPE,
  META_TYPE,
  STATUS_TYPE,
  STATUS_CHOICES,
  SELECTION_TYPE,
  ALL_SELECTED,
}
