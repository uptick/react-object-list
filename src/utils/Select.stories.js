import React from 'react'
import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'

import Select from './Select'

storiesOf('object-list/Utils/Select', module)
  .addDecorator((story, context) => withInfo(
    'Wrapper for react-select to ensure consistency'
  )(story)(context))
  .add('has custom menu style', () => (
    <Select menuStyle={{border: '5px dashed blue'}} />
  ))
  .add('has custom menu container style', () => (
    <Select menuContainerStyle={{border: '5px solid red'}} />
  ))
