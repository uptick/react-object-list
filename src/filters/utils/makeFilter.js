import React from 'react'
import PropTypes from 'prop-types'
import FilterLabel from './FilterLabel'
import FilterComparison from './FilterComparison'
import RemoveFilter from './RemoveFilter'
import {FILTER_BASE_TYPE} from '../../utils/proptypes'

/**
 * Utility function to make a filter class with
 * the provided Renderer for its value input
 * @param  {React.Component} Renderer Component used to render input
 * @return {React.Component}          Component used to render filter
 */
const makeFilter = (Renderer) => {
  /**
   * Base filter class
   * Renderers common label, comparison and remove components
   */
  class Filter extends React.Component {
    static propTypes = {
      ...FILTER_BASE_TYPE,
      ...Renderer.propTypes,
      /** Function to be called when filter is updated */
      onChange: PropTypes.func.isRequired,
      /** Function to be called when filter is removed */
      removeFilter: PropTypes.func.isRequired,
      /** Options to be passed to the comparison dropdown */
      comparisonOptions: FilterComparison?.propTypes?.options,
      /** Currently selected comparison option */
      comparison: FilterComparison?.propTypes?.value,
      /** Icons */
      icons: PropTypes.object,
      /** Object of custom react-select styles */
      selectStyles: PropTypes.object,
    }

    static defaultProps = {
      comparisonOptions: [],
      icons: {},
      permanent: false,
      selectStyles: {},
    }

    /**
     * Handler to call onChange with the newly selected comparison
     * along with the other filter options - only if it has changed
     * @param newComparison Comparison just selected
     */
    onComparisonChange = (newComparison) => {
      const {filterKey, value, comparison} = this.props
      if (newComparison !== comparison) {
        this.props.onChange({filter: filterKey, comparison: newComparison, value: value})
      }
    }

    /**
     * Handler to call onChange with the new value
     * along with other filter options - only if it has changed
     * @param newValue Value just entered
     */
    onValueChange = (newValue) => {
      const {filterKey, value, comparison} = this.props
      if (newValue !== value) {
        this.props.onChange({filter: filterKey, comparison: comparison, value: newValue})
      }
    }

    /**
     * Handler to call removeFilter with the filter key
     * @param  {MouseEvent} event onClick event
     */
    removeFilter = (event) => {
      const {removeFilter, filterKey} = this.props
      event.preventDefault()
      removeFilter(filterKey)
    }

    render() {
      const {
        name, comparison, comparisonOptions, value, permanent, selectStyles, icons = {}, ...otherProps
      } = this.props
      return (
        <div className="objectlist-row">
          {name && <FilterLabel label={name} />}
          {!!comparisonOptions.length && (
            <FilterComparison
              options={comparisonOptions}
              value={comparison || comparisonOptions[0]}
              onChange={this.onComparisonChange}
              selectStyles={selectStyles}
            />
          )}
          <Renderer
            {...otherProps}
            comparison={comparison || comparisonOptions[0]}
            value={value}
            onChange={this.onValueChange}
            icons={icons}
            selectStyles={selectStyles}
          />
          {!permanent && (
            <RemoveFilter onClick={this.removeFilter} Icon={icons.RemoveFilter} />
          )}
        </div>
      )
    }
  }
  return Filter
}

export default makeFilter
