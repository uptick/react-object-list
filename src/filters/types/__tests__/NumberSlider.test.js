import React from 'react'
import { shallow } from 'enzyme'
import { snapshotTest } from 'utils/tests'
import NumberSlider from '../NumberSlider'

describe('NumberSlider', () => {
  const baseProps = {
    onChange: jest.fn(),
    value: 50,
    precision: 1,
  }
  describe('Snapshots', () => {
    it('renders default', () => {
      snapshotTest(<NumberSlider {...baseProps} />)
    })
    it('custom min', () => {
      snapshotTest(<NumberSlider {...baseProps} min={30} />)
    })
    it('custom max', () => {
      snapshotTest(<NumberSlider {...baseProps} max={2000} />)
    })
  })
  describe('Lifecycle', () => {
    it('componentWillReceiveProps', () => {
      const newProps = {
        value: 3128907234589789,
      }
      const wrapper = shallow(<NumberSlider {...baseProps} />)
      wrapper.setProps(newProps)
      expect(wrapper.instance().state.currentValue).toEqual(newProps.value)
    })
  })
  describe('Functions', () => {
    let instance
    beforeEach(() => {
      spyOn(baseProps, 'onChange')
      instance = shallow(<NumberSlider {...baseProps} />).instance()
    })
    it('handles slider value change', () => {
      const mockValue = Math.floor(Math.random() * 100) + 1
      const mockEvent = {
        target: {
          value: mockValue,
        },
      }
      instance.handleSliderValueChange(mockEvent)
      expect(instance.state.currentValue).toBe(mockValue)
      expect(baseProps.onChange).not.toHaveBeenCalled()
    })
    it('handles value change', () => {
      const mockValue = Math.floor(Math.random() * 100) + 1
      const mockEvent = {
        target: {
          value: mockValue,
        },
      }
      instance.handleValueChange(mockEvent)
      expect(instance.state.currentValue).toBe(mockValue)
      expect(baseProps.onChange).toHaveBeenCalledWith(mockValue.toFixed(1))
    })
    it('handles slider value finalise', () => {
      const mockValue = Math.floor(Math.random() * 100) + 1
      const mockEvent = {
        target: {
          value: mockValue,
        },
      }
      instance.handleSliderValueFinalise(mockEvent)
      expect(instance.state.currentValue).toBe(mockValue)
      expect(baseProps.onChange).toHaveBeenCalledWith(mockValue.toFixed(1))
    })
    it('handles logarithmic translation', () => {
      const mockValue = Math.floor(Math.random() * 100) + 1
      instance = shallow(<NumberSlider {...baseProps} logarithmic />).instance()
      const actualValue = instance.getValueFromSlider(mockValue)
      expect(instance.getValueForSlider(actualValue)).toBeCloseTo(mockValue, 10)
    })
  })
})
