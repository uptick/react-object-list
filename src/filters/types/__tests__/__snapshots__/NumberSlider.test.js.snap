// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberSlider Snapshots custom max 1`] = `
<div
  className="objectlist-row objectlist-current-filter__active-status objectlist-number-slider-filter"
>
  <input
    className="objectlist-input objectlist-input__number-slider--number"
    max={2000}
    min={0}
    onChange={[Function]}
    step={0.1}
    type="number"
    value="50.0"
  />
  <input
    className="objectlist-input__number-slider--slider"
    max={2000}
    min={0}
    onChange={[Function]}
    onMouseUp={[Function]}
    step="any"
    type="range"
    value="50.0000000000"
  />
</div>
`;

exports[`NumberSlider Snapshots custom min 1`] = `
<div
  className="objectlist-row objectlist-current-filter__active-status objectlist-number-slider-filter"
>
  <input
    className="objectlist-input objectlist-input__number-slider--number"
    max={100}
    min={30}
    onChange={[Function]}
    step={0.1}
    type="number"
    value="50.0"
  />
  <input
    className="objectlist-input__number-slider--slider"
    max={100}
    min={30}
    onChange={[Function]}
    onMouseUp={[Function]}
    step="any"
    type="range"
    value="50.0000000000"
  />
</div>
`;

exports[`NumberSlider Snapshots renders default 1`] = `
<div
  className="objectlist-row objectlist-current-filter__active-status objectlist-number-slider-filter"
>
  <input
    className="objectlist-input objectlist-input__number-slider--number"
    max={100}
    min={0}
    onChange={[Function]}
    step={0.1}
    type="number"
    value="50.0"
  />
  <input
    className="objectlist-input__number-slider--slider"
    max={100}
    min={0}
    onChange={[Function]}
    onMouseUp={[Function]}
    step="any"
    type="range"
    value="50.0000000000"
  />
</div>
`;
